File: app/src/main/java/com/example/admin/smarttranslator/Services/PhotoService.java
package com.example.admin.smarttranslator.Services;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.support.media.ExifInterface;
import android.util.Log;

import java.io.FileOutputStream;
import java.io.IOException;

public class PhotoService {

    public Bitmap decodingPhoto(String path, int width, int height){
        BitmapFactory.Options options = new BitmapFactory.Options();
        options.inJustDecodeBounds = true;
        BitmapFactory.decodeFile(path, options);
        options.inSampleSize = calculateInSampleSize(options, width, height);
        options.inJustDecodeBounds = false;
        Bitmap resultBitmap = BitmapFactory.decodeFile(path, options);

        resultBitmap = changeOrientation( resultBitmap, getExifAngle(path) );
        return resultBitmap;
    }

    public void compressPhoto(Bitmap bitmap, String path) {

        try(FileOutputStream out = new FileOutputStream(path)){

            bitmap.compress(Bitmap.CompressFormat.JPEG,20,out);

        } catch (IOException e) {
            e.printStackTrace();
        }

    }

    private int calculateInSampleSize(BitmapFactory.Options options, int reqWidth, int reqHeight) {
        final int height = options.outHeight;
        final int width = options.outWidth;
        int inSampleSize = 1;

        if (height > reqHeight || width > reqWidth) {

            final int halfHeight = height / 2;
            final int halfWidth = width / 2;

             while ((halfHeight / inSampleSize) >= reqHeight && (halfWidth / inSampleSize) >= reqWidth) {
                inSampleSize *= 2;
            }
        }

        return inSampleSize;
    }

    private Bitmap changeOrientation(Bitmap bitmap, int angle){
        Matrix matrix = new Matrix();
        matrix.postRotate(angle);
        return Bitmap.createBitmap(bitmap,0,0,bitmap.getWidth(),bitmap.getHeight(),matrix,true);
    }

    private static int getExifAngle(String path) {
        int angle = 0;
        try {
            ExifInterface ei = new ExifInterface(path);
            int orientation = ei.getAttributeInt(ExifInterface.TAG_ORIENTATION, ExifInterface.ORIENTATION_UNDEFINED);
            switch(orientation) {
                case ExifInterface.ORIENTATION_ROTATE_90:
                    angle = 90;
                    break;
                case ExifInterface.ORIENTATION_ROTATE_180:
                    angle = 180;
                    break;
                case ExifInterface.ORIENTATION_ROTATE_270:
                    angle = 270;
                    break;

                default:angle = 0;
                    break;
            }
        } catch (Exception e) {
            Log.d("!!!!!!_ANGLE", e.toString());
        }
        return angle;
    }

}

File: app/src/main/java/com/example/admin/smarttranslator/Services/PhotoService.kt
package com.example.admin.smarttranslator.Services

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Matrix
import androidx.exifinterface.media.ExifInterface
import android.util.Log

import java.io.FileOutputStream
import java.io.IOException

class PhotoService {

    fun decodingPhoto(path: String, width: Int, height: Int): Bitmap {
        val options = BitmapFactory.Options()
        options.inJustDecodeBounds = true
        BitmapFactory.decodeFile(path, options)
        options.inSampleSize = calculateInSampleSize(options, width, height)
        options.inJustDecodeBounds = false
        var resultBitmap = BitmapFactory.decodeFile(path, options)

        resultBitmap = changeOrientation(resultBitmap, getExifAngle(path))
        return resultBitmap
    }

    fun compressPhoto(bitmap: Bitmap, path: String) {

        try {
            FileOutputStream(path).use { out ->

                bitmap.compress(Bitmap.CompressFormat.JPEG, 20, out)

            }
        } catch (e: IOException) {
            e.printStackTrace()
        }

    }

    private fun calculateInSampleSize(options: BitmapFactory.Options, reqWidth: Int, reqHeight: Int): Int {
        val height = options.outHeight
        val width = options.outWidth
        var inSampleSize = 1

        if (height > reqHeight || width > reqWidth) {

            val halfHeight = height / 2
            val halfWidth = width / 2

            while (halfHeight / inSampleSize >= reqHeight && halfWidth / inSampleSize >= reqWidth) {
                inSampleSize *= 2
            }
        }

        return inSampleSize
    }

    private fun changeOrientation(bitmap: Bitmap, angle: Int): Bitmap {
        val matrix = Matrix()
        matrix.postRotate(angle.toFloat())
        return Bitmap.createBitmap(bitmap, 0, 0, bitmap.width, bitmap.height, matrix, true)
    }

    private fun getExifAngle(path: String): Int {
        var angle = 0
        try {
            val ei = androidx.exifinterface.media.ExifInterface(path)
            val orientation = ei.getAttributeInt(androidx.exifinterface.media.ExifInterface.TAG_ORIENTATION, androidx.exifinterface.media.ExifInterface.ORIENTATION_UNDEFINED)
            when (orientation) {
                androidx.exifinterface.media.ExifInterface.ORIENTATION_ROTATE_90 -> angle = 90
                androidx.exifinterface.media.ExifInterface.ORIENTATION_ROTATE_180 -> angle = 180
                androidx.exifinterface.media.ExifInterface.ORIENTATION_ROTATE_270 -> angle = 270

                else -> angle = 0
            }
        } catch (e: Exception) {
            Log.d("!!!!!!_ANGLE", e.toString())
        }

        return angle
    }

}
